### PR TITLE
Add REST API Filters to modify API responses

### PR DESCRIPTION
**Pull request by moon0326**
This PR adds a new tab that allows us to modify API responses using dot notation without changing code. It can come in handy when testing a new change that depends on an API response. 

### Testing Instruction

1. Install WooCommerce
2. Navigate to `WooComemrce -> Home` and click `Set up payments` task. We're going to modify the response of `/wc/v3/payment_gateways` endpoint.
3. Open inspector -> Network and select `payment_gateways?_locale=user`
4. Click `Preview` tab and confirm the first array element has `needs_setup` field and its value is `true`
5. Navigate to `Tools -> WCA Test Helper -> REST API Filters`
6. Click `New Filter` and enter the following values. Click `Create New Filter` button.

**Endpoint:** /wc/v3/payment_gateways
**Dot Notation:** 0.needs_setup
**Replacement:** false

7. Go back to the payment task and refresh the page.
8. Check the response again and confirm the `needs_setup` field is set to false.

### Screenshots

Screen Shot 2022-04-14 at 2 01 29 PM

### CODE REVIEW

File: api/rest-api-filters/hook.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+add_filter(
+    'rest_request_after_callbacks',
+    function( $response, array $handler, \WP_REST_Request $request ) use ( $filters ) {
+		if ( ! $response instanceof  \WP_REST_Response ) {
+			return $response;
+		}
+	    $route = $request->get_route();
+		$filters = array_filter( $filters, function( $filter ) use ( $request, $route ) {
```
Comment (chihsuan) [inline api/rest-api-filters/hook.php:31]: @moon0326 Looks like our PHP code isn't well-formatted. Do we need to install [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer)?

File: api/rest-api-filters/rest-api-filters.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+class WCA_Test_Helper_Rest_Api_Filters {
+	const WC_ADMIN_TEST_HELPER_REST_API_FILTER_OPTION = 'wc-admin-test-helper-rest-api-filters';
+
+	public static function create( $request ) {
+		$endpoint = $request->get_param( 'endpoint' );
+		$dot_notation = $request->get_param( 'dot_notation' );
+		$replacement = $request->get_param( 'replacement' );
+
+		if ( $replacement === 'false' ) {
+			$replacement = false;
```
Comment (chihsuan) [inline api/rest-api-filters/rest-api-filters.php:65]: It seems that we could use `filter_var( $replacement, FILTER_VALIDATE_BOOLEAN );`?

https://stackoverflow.com/questions/4775294/parsing-a-string-into-a-boolean-value-in-php

**Review by chihsuan — CHANGES_REQUESTED** [review]
Tested well! I just have some suggestions and questions.

**Comment by chihsuan** [comment]
Screen Shot 2022-04-15 at 11 10 15

```
Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40 Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40 Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40

```

@moon0326 I also got this warning when I used a fresh site.

**Comment by moon0326** [comment]
> Screen Shot 2022-04-15 at 11 10 15
> 
> ```
> Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40 Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40 Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40
> ```
> 
> @moon0326 I also got this warning when I used a fresh site.

Thank you for the review! I'll address them soon 👍

File: api/rest-api-filters/hook.php (hunk 1 above)
Comment (moon0326) [reply api/rest-api-filters/hook.php]: Fixed in https://github.com/woocommerce/woocommerce-admin-test-helper/pull/37/commits/c3ad996eb28adba93a51fb9a5bc0a37925eee94f.

I'll work a follow up PR to add PHP linters

File: src/rest-api-filters/index.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+			// eslint-disable-next-line camelcase
+			const dotNotation = dot_notation;
+			return (
+				<tr key={ index }>
+					<td key={ 0 }>{ index + 1 }</td>
+					<td key={ 1 }>{ endpoint }</td>
+					<td key={ 'optionValue' }>{ dotNotation }</td>
+					<td className="align-center" key={ 2 }>
+						{ replacement + '' }
```
[the remark this answers is not in the document]
Comment (moon0326) [reply src/rest-api-filters/index.js:72]: So I noticed that React doesn't render true or false. Adding an empty string force type conversation to render the value.

File: src/rest-api-filters/index.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+		return filters.map( ( filter, index ) => {
+			// eslint-disable-next-line camelcase
+			const { endpoint, dot_notation, replacement, enabled } = filter;
+
+			// eslint-disable-next-line camelcase
+			const dotNotation = dot_notation;
+			return (
+				<tr key={ index }>
+					<td key={ 0 }>{ index + 1 }</td>
```
[the remark this answers is not in the document]
Comment (moon0326) [reply src/rest-api-filters/index.js]: Fixed in https://github.com/woocommerce/woocommerce-admin-test-helper/pull/37/commits/a9f2322d8cdc95ea1195b315194e0c697bc98d4d

File: src/rest-api-filters/index.js [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+					onClick={ () => setNewModalOpen( true ) }
+				/>
+				<br />
+				<br />
+				<table className="wp-list-table striped table-view-list widefat">
+					<thead>
+						<tr>
+							<td
+								className="manage-column column-thumb"
+								key={ 0 }
```
[the remark this answers is not in the document]
Comment (moon0326) [reply src/rest-api-filters/index.js]: Fixed in https://github.com/woocommerce/woocommerce-admin-test-helper/pull/37/commits/a9f2322d8cdc95ea1195b315194e0c697bc98d4d

File: api/rest-api-filters/rest-api-filters.php (hunk 2 above)
Comment (moon0326) [reply api/rest-api-filters/rest-api-filters.php]: Nice! I did not know about that :) 

Fixed in https://github.com/woocommerce/woocommerce-admin-test-helper/pull/37/commits/f0e8ad9c86a4ee5c40e081f9e27c86c05bf50f38

**Comment by moon0326** [comment]
> Screen Shot 2022-04-15 at 11 10 15
> 
> ```
> Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40 Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40 Warning: array_filter() expects parameter 1 to be array, bool given in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 36 Warning: Invalid argument supplied for foreach() in /var/www/html/wp-content/plugins/woocommerce-admin-test-helper/api/rest-api-filters/hook.php on line 40
> ```
> 
> @moon0326 I also got this warning when I used a fresh site.

Fixed in https://github.com/woocommerce/woocommerce-admin-test-helper/pull/37/commits/db8ccabb15c268fa27d31e88fd0aad9c77d859be

File: src/rest-api-filters/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (chihsuan) [reply src/rest-api-filters/index.js]: Got it! I think using explicit type conversation like `String()` or `.toString()` might be better, but I'm good with this. 🙂

**Review by chihsuan — APPROVED** [review]
@moon0326 Thanks for adding this useful feature. Looks good and tested well. 👍

**Pull request merged**